### PR TITLE
ensure cli config has higher priority than config files

### PR DESCRIPTION
**Pull request by minrk**

### CODE REVIEW

File: jupyter_core/tests/test_application.py [hunk 1]
```diff
@@ -1,12 +1,17 @@
+import os
 from tempfile import NamedTemporaryFile
 
 from traitlets import Integer
 
 from jupyter_core.application import JupyterApp
 
+pjoin = os.path.join
```
Comment (damianavila) [inline jupyter_core/tests/test_application.py]: but `pjoin` is not used anywhere here (also the `import os` would not be necessary)... maybe a leftover of a previous implementation or I am missing something?

Comment (minrk) [inline jupyter_core/tests/test_application.py]: artefact leftover from a rebase. Removed.

Comment (takluyver) [inline jupyter_core/tests/test_application.py]: Ditto the `import os` above. And I think you forgot to push?

Comment (minrk) [inline jupyter_core/tests/test_application.py]: pushed, just to the wrong repo

File: jupyter_core/tests/test_application.py [hunk 2]
```diff
@@ -27,3 +29,12 @@ def test_custom_config():
         app.initialize(['--config', f.name])
     assert app.config_file == f.name
     assert app.n == 10
+
+
+def test_cli_override():
+    app = DummyApp()
+    with NamedTemporaryFile(suffix='.py', mode='w') as f:
```
Comment (takluyver) [inline jupyter_core/tests/test_application.py]: Will this work on Windows? It needs to open two handles on the file at the same time, and I seem to remember that's problematic. We have `NamedFileInTemporaryDirectory` in genutils for this.

Comment (minrk) [inline jupyter_core/tests/test_application.py]: I'll just use mkdtemp to be safe.

**Pull request merged**